sound/soc/sof/sof-audio.c

```diff
@@ -198,65 +271,240 @@ static int sof_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swi
+	ret = sof_ipc_tx_message(sdev->ipc, connect->hdr.cmd, connect, sizeof(*connect),
+				 &reply, sizeof(reply));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to load route source %s -> %s sink\n",
+			wsource->name, wsink->name);
+	} else {
+		dev_dbg(sdev->dev, "route %s -> %s setup complete\n", wsource->name, wsink->name);
+		sroute->setup = true;
+	}
 
-		ret = sof_dai_config_setup(sdev, dai);
-		if (ret < 0) {
-			dev_err(sdev->dev, "error: failed to load dai config for DAI %s\n",
-				swidget->widget->name);
-			return ret;
+	return ret;
+}
+
+static bool is_sof_widget(struct snd_soc_dapm_widget *widget)
+{
+	switch (widget->id) {
+	case snd_soc_dapm_dai_in:
+	case snd_soc_dapm_dai_out:
+	case snd_soc_dapm_mixer:
+	case snd_soc_dapm_pga:
+	case snd_soc_dapm_buffer:
+	case snd_soc_dapm_scheduler:
+	case snd_soc_dapm_aif_out:
+	case snd_soc_dapm_aif_in:
+	case snd_soc_dapm_src:
+	case snd_soc_dapm_asrc:
+	case snd_soc_dapm_siggen:
+	case snd_soc_dapm_effect:
+	case snd_soc_dapm_mux:
+	case snd_soc_dapm_demux:
+		if (widget->dobj.private)
+			return true;
+
+		return false;
+	default:
+		return false;
+	}
+}
+
+static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
+					  struct snd_soc_dapm_widget_list *list, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int i;
+
+	/*
+	 * Set up connections between widgets in the sink/source paths based on direction.
+	 * Some non-SOF widgets exist in topology either for compatibility or for the
+	 * purpose of connecting a pipeline from a host to a DAI in order to receive the DAPM
+	 * events. But they are not handled by the firmware. So ignore them.
+	 */
+	if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_sink_path(widget, p)
+				if (is_sof_widget(p->sink)) {
+					ret = sof_route_setup(sdev, widget, p->sink);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	} else {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_source_path(widget, p)
+				if (is_sof_widget(p->source)) {
+					ret = sof_route_setup(sdev, p->source, widget);
+					if (ret < 0)
+						return ret;
+				}
 		}
 	}
 
 	return 0;
 }
 
-/* placeholder for now. Will be updated in the following patches */
 int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
 {
 	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_soc_dapm_widget *widget;
+	int num_pipe_widgets = 0;
+	int i, ret, num_widgets;
 
 	/* nothing to set up */
 	if (!list)
 		return 0;
 
-	return sof_widget_list_dai_config_setup(sdev, list);
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, num_widgets, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
```

Comment:
if two pcm share one pipeline like mix case, these widgets in the pipeline will be setup twice ?

Reply:
also shared pipeline

Reply:
@RanderWang no, if there's mixer it will part of a separate pipeline. Have a look at the BSW/BYT topology. And no they will not be setup twice, sof_widget_setup() would simply increment the use_count and return. And when the PCM is freed, sof_widget_free() would decrement the use_count and free when the use_count reaches 0.

Reply:
yes, I recall it when I woke up in the morning.